src/ir_passes/mono/monomorphize.rs

```diff
@@ -0,0 +1,163 @@
+                entrypoint: None,
+                externals: HashMap::new(),
+            },
+            old,
+            externals: vec![],
+            processed: HashMap::new(),
+            queue: LinkedHashMap::new(),
+            port_map: HashMap::new(),
+            event_map: HashMap::new(),
+            ext_map: HashMap::new(),
+        }
+    }
+}
+
+impl<'ctx> Monomorphize<'ctx> {
+    /// Queue an instance for processing by the pass.
+    /// The processing happens at a later point but, if needed, the pass immediately allocates a new [ir::Component] and returns information to construct a new instance.
+    pub fn should_process(
+        &mut self,
+        comp: ir::CompIdx,
+        params: Vec<u64>,
+    ) -> (ir::CompIdx, Vec<u64>) {
+        let underlying = self.old.get(comp);
+
+        // If it is an external, add it to externals
+        if underlying.is_ext {
+            self.externals.push(comp);
+        }
+
+        let key = (comp, params.clone());
+
+        // If we've already processed this or queued this for processing, return the component
+        if let Some(&name) = self.processed.get(&key) {
+            return (name, vec![]);
+        }
+
+        if let Some((name, _)) = self.queue.get(&key) {
+            return (*name, vec![]);
+        }
+
+        // Otherwise, construct a new component and add it to the processing queue
+        let new_comp = self.ctx.comp(underlying.is_ext, &underlying.filename);
+        if underlying.filename.is_some() {
+            let filename = underlying.filename.clone().unwrap();
+            if let Some(exts) = self.ext_map.get(&filename) {
+                let mut exts = exts.clone();
+                exts.push(new_comp);
+                self.ext_map.insert(filename, exts);
+            } else {
+                self.ext_map.insert(filename, vec![new_comp]);
+            }
+        }
+
+        let base = self.ctx.get_mut(new_comp);
+
+        // make a MonoSig
+        let mut monosig = MonoSig::new(base, underlying, comp, params);
+
+        // the component whose signature we want to monomorphize
+        let underlying = self.old.get(comp);
+
+        // Monomorphize the sig
+        MonoDeferred::sig(&mut monosig, underlying, self);
+
+        // Insert into queue, with monosig so we can pick up where we left off when ready
+        self.queue.insert(key, (new_comp, monosig));
+
+        // return the `base` index so we can update the instance
+        (new_comp, vec![])
+    }
+
+    fn next(&mut self) -> Option<(ir::Component, ir::CompIdx)> {
+        let Some(((underlying_idx, params), (base_idx, monosig))) = self.queue.pop_front() else {
+            return None;
+        };
+
+        self.processed.insert((underlying_idx, params), base_idx);
+
+        let underlying = self.old.get(underlying_idx);
+        let mut mono = MonoDeferred {
+            underlying,
+            pass: self,
+            monosig,
+        };
+
+        mono.gen_comp();
```

Comment:
Make `gen_comp` return the information used below.

Reply:
Oh weird okay! Does `gen_comp` take `&mut self` or `self`? The latter should allow returning it but no worries if not!

Reply:
Yeah it takes `&mut self`